packages/viz/src/utils/print/data.js

```diff
@@ -178,101 +44,261 @@ function processBgRange(selectedDataByDate) {
-
+function processBolusRange(dataByDate) {
   const boluses = _.reduce(
-    selectedDataByDate, (all, date) => (all.concat(_.get(date, "data.bolus", []))), []
+    dataByDate,
+    (all, date) => (
+      all.concat(_.get(date, "data.bolus", []))
+    ),
+    []
   );
-  _.forEach(boluses, (bolus) => {
-    // eslint-disable-next-line no-param-reassign
-    bolus.threeHrBin = Math.floor(moment.utc(bolus.utc).tz(selected.timezone).hours() / 3) * 3;
-  });
-  selected.bolusRange = extent(boluses, (d) => (d.normal + (d.extended || 0)));
-
-  _.forEach(selectedDataByDate, (dateObj) => {
-    const { data: { bolus: bolusesForDate } } = dateObj;
-    selectedDataByDate[dateObj.date].data.bolus = _.map(bolusesForDate, (bolus) => {
-      if (bolus.wizard) {
-        const reversed = stripDatum(bolus.wizard);
-        reversed.bolus = _.omit(bolus, "wizard");
-        return reversed;
-      }
-      return bolus;
-    });
+  return extent(boluses, (d) => {
+    const bolus = d.bolus ? d.bolus : d;
+    return bolus.normal + (bolus.extended ?? 0);
   });
+}
 
-  const allBasals = _.reduce(
-    selectedDataByDate, (all, date) => (all.concat(_.get(date, "data.basal", []))), []
+function processBasalRange(dataByDate) {
+  const basals = _.reduce(
+    dataByDate,
+    (all, date) => (
+      all.concat(_.get(date, "data.basal", []))
+    ),
+    []
   );
   const rawBasalRange = extent(
-    allBasals,
+    basals,
     (d) => (_.max([_.get(d, "suppressed.rate", 0), d.rate]))
   );
   // multiply the max rate by 1.1 to add a little buffer so the highest basals
   // don't sit at the very top of the basal rendering area and bump into boluses
-  selected.basalRange = [0, rawBasalRange[1] * 1.1];
-
-  _.forEach(selected.dataByDate, (dateData) => {
-    const { bounds, data: { basal: basals } } = dateData;
-    for (let i = 0; i < basals.length; ++i) {
-      const basal = basals[i];
-      // trim the first and last basals to fit within the date's bounds
-      if (basal.utc < bounds[0]) {
-        basal.duration = basal.duration - (bounds[0] - basal.utc);
-        basal.utc = bounds[0];
-      }
-      if (i === basals.length - 1 && dateData.date !== selected.dateRange[1]) {
-        basal.duration = bounds[1] - basal.utc;
-      }
-      let nextBasal;
-      basal.subType = basal.deliveryType;
-      delete basal.deliveryType;
-      if (i !== basals.length - 1) {
-        nextBasal = basals[i + 1];
-        if ((basal.utc + basal.duration) !== nextBasal.utc) {
-          basal.discontinuousEnd = true;
-          nextBasal.discontinuousStart = true;
+  return [0, rawBasalRange[1] * 1.1];
+}
+
+/**
+ *
+ * @param {{duration:number;epoch:number;epochEnd:number;subType:string;discontinuousEnd:boolean;discontinuousStart:boolean;}[]} basals
+ * @param {[number,number]} bounds
+ */
+function updateBasalDiscontinuous(basals, bounds) {
+  if (basals.length < 1) {
+    return;
+  }
+  let prevBasal = null;
+  for (let i=0; i<basals.length; i++) {
```

Comment:
```suggestion
  for (let basal of basals) {
```

Reply:
in this case, given the number of elements we may have, I'm not sure it's a good idea.
After some testing between the 2 implementations (complete days loops), with 3 months of data:

Tests run:
| Impl  | 1 | 2 | 3 | 4 | 5 | Avg |
| ------ | --- | --- | --- | --- | --- | --- |
| `i++` | 1271 ms | 1257 ms | 1430 ms | 1261 ms | 1406 ms | 1325 ms |
| `of`   | 1428 ms | 1390 ms | 1811 ms | 1420 ms | 1791 ms | 1568 ms |

For this specific case, I don't think it's worth it.